### ISSUE TITLE
curses.setupterm must only be called once per process

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3269]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3269|
|Type|defect|
|Created|2008-06-02 09:03:42Z|
|Branch|https://github.com/twisted/twisted/tree/curses-setupterm-3269|

The class _AnsiColorizer in twisted.trial.reporter calls it for each supported() call. But this function seems to leak memory, so it should only be called if necessary. Unfortunately, there is no way to check if it has been called expect by trying to call a curses function needing it.

<details><summary>Searchable metadata</summary>

```
trac-id__3269 3269
type__defect defect
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_curses_setupterm_3269 branches/curses-setupterm-3269
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1212397422000000 1212397422000000
changetime__1214830364000000 1214830364000000
version__None None
owner__therve therve

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [23812]) Branching to 'curses-setupterm-3269'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

`os.environ` is really terrible.  I think the tests need to do something slightly different with respect to it.  Consider:

```
>>> import os
>>> os.system('echo $LESSOPEN')
| /usr/bin/lesspipe %s
0
>>> os.environ.pop('LESSOPEN')
'| /usr/bin/lesspipe %s'
>>> os.system('echo $LESSOPEN')
| /usr/bin/lesspipe %s
0
>>> os.environ['LESSOPEN'] = 'hello, world'
>>> os.system('echo $LESSOPEN')
hello, world
0
>>> os.environ = {'LESSOPEN': 'foo bar'}
>>> os.system('echo $LESSOPEN')
hello, world
0
>>>
```

Also, in `setUp`, I think you probably meant to save `os.environ.copy()` instead of just `os.environ`, but I don't think that helps at all.

I feel like `_AnsiColorizer` should be using a terminfo database instead of `curses` (or it should be using `curses` to do the coloring as well, not producing color codes on its own).  That may be a different ticket, though.  I don't know if there's even a good way to read terminfo from Python, anyway.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [23847]) Remove cygwin related code.

Refs [#3269](https://github.com/twisted/twisted/issues/3269)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

I removed the code related to cygwin. This is not a supported platform, and I have no idea on how that works (I tried using cygwin with a win32 python installed, and that worked flawlessly). Hopefully someone interested later could add tests for that.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

Changes look good.  I noticed a pre-existing behavior which may be a defect.  `_AnsiColorizer` checks `sys.stderr.isatty()` but the reporter writes to `sys.stdout`.  This means `trial foo > bar` will probably write colors to `bar`.  I doubt this is intentional or desirable.  I rather suspect that `supported` should accept the file-like object which is going to be written to (`__init__` already does) and make the check against that, instead.

I'd like to take another look if you fix the stdout/stderr issue in this branch.  If you'd rather file a separate ticket for that, cool and go ahead and merge this.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve removed owner|
|-|-|

Back to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

  * `_Win32Colorizer` still checks stdout, even if it is passed something else.  I don't really care much about this.  I don't even know how you'd get trial to be writing to something other than stdout on Windows.  You could file another ticket for it, if you want.
  * `twisted.trial.test.test_reporter.AnsiColorizerTests.test_supportedSetupTerm` and `twisted.trial.test.test_reporter.AnsiColorizerTests.test_supportedTigetNumErrors` fail if run when stdout is not a pty.
  * `MockColorizer.supported` has a signature incompatible with the other colorizers.  Since nothing fails because of this, I suppose that means it is never called, so maybe we don't need it

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Points 2 and 3 addressed. I don't really care about 1 either.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Looks good.  Merge if http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/curses-setupterm-3269&num_builds=1 goes green.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [24169]) Merge curses-setupterm-3269 

Author: therve
Reviewer: exarkun
Fixes [#3269](https://github.com/twisted/twisted/issues/3269)

Fix trial AnsiColorizer so that it doesn't call curses.setupterm uselessly: it
caused a (small) memory leak in the reporter when using trial --until-failure.
